### PR TITLE
Pass source origin in all fetch requests

### LINKED ISSUE #1762: Send AMP-Origin header with fetch requests
Recipients need to check this as well as the standard origin header.

### PR DESCRIPTION
**Pull request by dvoytenko**
Closes #1762.

### CODE REVIEW

**Comment by erwinmombay** [comment]
@dvoytenko looks good to me, passing onto @cramforce

File: src/xhr.js [hunk 1]
```diff
@@ -59,6 +67,29 @@ class Xhr {
...
+   */
+  performFetch_(input, opt_init) {
+    // Add "__amp_source_origin" query parameter to the URL. Ideally, we'd be
+    // able to set a header (e.g. AMP-Source-Origin), but this will force
+    // preflight request on all CORS request.
+    const sourceOrigin = getSourceOrigin(this.win.location.href);
+    const url = parseUrl(input);
+    const queryString = parseQueryString(url.search);
+    const querySourceOrigin = queryString[SOURCE_ORIGIN_PARAM];
+    if (!querySourceOrigin) {
```
Comment (cramforce) [inline src/xhr.js]: if (SOURCE_ORIGIN_PARAM in queryString) {
  throw …
}

File: src/xhr.js [hunk 2]
```diff
@@ -59,6 +67,29 @@ class Xhr {
...
+   * @param {string} input
+   * @param {?FetchInitDef=} opt_init
+   * @return {!Promise<!FetchResponse>}
+   * @private
+   */
+  performFetch_(input, opt_init) {
+    // Add "__amp_source_origin" query parameter to the URL. Ideally, we'd be
+    // able to set a header (e.g. AMP-Source-Origin), but this will force
+    // preflight request on all CORS request.
+    const sourceOrigin = getSourceOrigin(this.win.location.href);
```
Comment (cramforce) [inline src/xhr.js]: Do an additional sanity check:

`if (decodeURI(input).indexOf('__amp_source_origin') != -1) throw…`

File: test/functional/test-xhr.js [hunk 3]
```diff
@@ -105,6 +117,27 @@ describe('XHR', function() {
...
+        setupMockXhr();
+        xhr.fetchJson('/get');
+        expect(noOrigin(requests[0].url)).to.equal(
+            '/get?__amp_source_origin=https%3A%2F%2Facme.com');
+      });
+
+      it('should defend against invalid source origin query parameter', () => {
+        setupMockXhr();
+        expect(() => {
+          xhr.fetchJson('/get?k=v1&__amp_source_origin=invalid#h1');
```
Comment (cramforce) [inline test/functional/test-xhr.js]: Also add test case where __amp_source_origin appears not in query string and for both where part of it is percent encoded.

File: src/xhr.js (hunk 1 above)
Comment (dvoytenko) [inline src/xhr.js]: Done

File: src/xhr.js (hunk 2 above)
Comment (dvoytenko) [inline src/xhr.js]: Done

File: test/functional/test-xhr.js (hunk 3 above)
Comment (dvoytenko) [inline test/functional/test-xhr.js]: Done

**Comment by dvoytenko** [comment]
@cramforce I've addressed the comments. And I also added the check for `AMP-Access-Control-Allow-Source-Origin`. PTAL.

File: src/xhr.js [hunk 4]
```diff
@@ -59,6 +70,47 @@ class Xhr {
...
+    const url = parseUrl(input);
+    const query = parseQueryString(url.search);
+    if (SOURCE_ORIGIN_PARAM in query ||
+            decodeURI(input).indexOf(SOURCE_ORIGIN_PARAM) != -1) {
```
Comment (cramforce) [inline src/xhr.js]: As discussed, remove this.

File: src/xhr.js [hunk 5]
```diff
@@ -59,6 +70,47 @@ class Xhr {
   }
 
   /**
+   * Performs the final initialization and requests the fetche.
+   * @param {string} input
+   * @param {?FetchInitDef=} opt_init
+   * @return {!Promise<!FetchResponse>}
+   * @private
+   */
+  performFetch_(input, opt_init) {
```
Comment (cramforce) [inline src/xhr.js]: `fetchAmpCors` also document

**Comment by cramforce** [comment]
LGTM with one comment

File: src/xhr.js (hunk 4 above)
Comment (dvoytenko) [inline src/xhr.js]: Done

File: src/xhr.js (hunk 5 above)
Comment (dvoytenko) [inline src/xhr.js]: Done

**Pull request merged**